File: 백준/Gold/1520. 내리막 길/내리막 길.py
import sys
input = sys.stdin.readline
# 시간초과 한 코드
# def dfs(r, c, value):
#     global answer
#     if r == N-1 and c == N-1:
#         answer += 1
#         return
#
#     for dr, dc in [(0,1),(0,-1),(1,0),(-1,0)]:
#         newR = dr+r
#         newC = dc+c
#         if 0<=newR<N and 0<=newC<M and not visited[newR][newC] and lst[newR][newC] < value:
#             visited[newR][newC] = 1
#             dfs(newR, newC, lst[newR][newC])
#             visited[newR][newC] = 0
#
#
# N, M = map(int, input().split())
# lst = [list(map(int, input().split())) for _ in range(N)]
# visited = [[0]*M for _ in range(N)]
# answer = 0
# dfs(0,0,lst[0][0])
# print(answer)

# DFS + DP 꼭 다시 풀어보기....
def dfs(r, c):
    if r == N-1 and c == M-1:
        return 1

    if dp[r][c] == -1:
        dp[r][c] = 0

        for dr, dc in [(0,1),(0,-1),(1,0),(-1,0)]:
            newR = dr+r
            newC = dc+c
            if 0<=newR<N and 0<=newC<M and lst[newR][newC] < lst[r][c]:
                dp[r][c] += dfs(newR,newC)

    return dp[r][c]


N, M = map(int, input().split())
lst = [list(map(int, input().split())) for _ in range(N)]
dp = [[-1]*M for _ in range(N)]

print(dfs(0,0))

File: 백준/Gold/1520. 내리막 길/내리막 길.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int N, M;
    static int[][] board;
    static long[][] dp;
    static int[] dr = {0, 0, -1, 1};
    static int[] dc = {-1, 1, 0, 0};
    public static void main(String[] args) throws Exception{
        init();
        solve();
    }

    private static void solve() {
        System.out.println(dfs(0, 0));
    }

    private static long dfs(int r, int c) {
        if (r == N-1 && c == M-1) {
            return 1;
        }

        if (dp[r][c] != -1) {
            return dp[r][c];
        }

        dp[r][c] = 0;

        for (int i = 0; i < 4; i++) {
            int newR = r + dr[i];
            int newC = c + dc[i];

            if (0 > newR || newR >= N || 0 > newC || newC >= M) continue;

            if (board[r][c] > board[newR][newC]) {
                dp[r][c] += dfs(newR, newC);
            }
        }
        return dp[r][c];
    }

    private static void init() throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        board = new int[N][M];
        dp = new long[N][M];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            Arrays.fill(dp[i], -1);
            for (int j = 0; j < M; j++) {
                int num = Integer.parseInt(st.nextToken());
                board[i][j] = num;
            }
        }
    }
}
